src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.InlineContext.cs

```diff
@@ -0,0 +1,601 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.InlineMethod
+{
+    internal abstract partial class AbstractInlineMethodRefactoringProvider<TInvocationSyntax, TExpressionSyntax, TMethodDeclarationSyntax, TStatementSyntax>
+    {
+        private readonly struct InlineMethodContext
+        {
+            /// <summary>
+            /// Statements should be inserted before the <see cref="StatementContainingCallee"/>.
+            /// </summary>
+            public ImmutableArray<SyntaxNode> StatementsToInsertBeforeCallee { get; }
+
+            /// <summary>
+            /// Statement containing the callee. All the generated declarations should be put before this node.
+            /// Example:
+            /// void Caller()
+            /// {
+            ///     var x = Callee(Foo());
+            /// }
+            /// void Callee(int i) { DoSomething(); }
+            /// LocalDeclarationSyntaxNode of x will the <see cref="StatementContainingCallee"/>.
+            /// And if there is any statements needs inserted, it needs to be inserted before this node.
+            /// </summary>
+            public SyntaxNode StatementContainingCallee { get; }
+
+            /// <summary>
+            /// Inline content for the callee method. It should replace <see cref="SyntaxNodeToReplace"/>.
+            /// </summary>
+            public SyntaxNode InlineSyntaxNode { get; }
```

Comment:
```suggestion
            public TExpressionSyntax InlineSyntaxNode { get; }
```

Reply:
Ah, it is not guaranteed to be Expression. 
In some sceanario it could also be statement syntax.
Like the example:
`void Caller() { Callee(); }`
`int Callee() { return 1;}`
It would be `int temp = 1;` and the `SyntaxNodeToReplace` would be the `Callee();`. 
And after inlining it will be
`void Caller() { int temp =1; }`
In most cases it would an expression to replace the invocationExpression

Reply:
you might be able to simplify this greatly by *always* having it be an expression.  And tehn at the *callsite* you decide waht to do depending on what sort of statement+invocation you have.